alibi_detect/od/pytorch/lof.py

```diff
@@ -0,0 +1,164 @@
+            ensembler: Optional[Ensembler] = None,
+            device: Optional[Union[Literal['cuda', 'gpu', 'cpu'], 'torch.device']] = None,
+            ):
+        """PyTorch backend for LOF detector.
+
+        Parameters
+        ----------
+        k
+            Number of nearest neighbors used to compute the local outlier factor. `k` can be a single
+            value or an array of integers. If `k` is a single value the score method uses the
+            distance/kernel similarity to the `k`-th nearest neighbor. If `k` is a list then it uses
+            the distance/kernel similarity to each of the specified `k` neighbors.
+        kernel
+            If a kernel is specified then instead of using `torch.cdist` the kernel defines the `k` nearest
+            neighbor distance.
+        ensembler
+            If `k` is an array of integers then the ensembler must not be ``None``. Should be an instance
+            of :py:obj:`alibi_detect.od.pytorch.ensemble.ensembler`. Responsible for combining
+            multiple scores into a single score.
+        device
+            Device type used. The default tries to use the GPU and falls back on CPU if needed.
+            Can be specified by passing either ``'cuda'``, ``'gpu'``, ``'cpu'`` or an instance of
+            ``torch.device``.
+        """
+        TorchOutlierDetector.__init__(self, device=device)
+        self.kernel = kernel
+        self.ensemble = isinstance(k, (np.ndarray, list, tuple))
+        self.ks = torch.tensor(k) if self.ensemble else torch.tensor([k], device=self.device)
+        self.ensembler = ensembler
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Detect if `x` is an outlier.
+
+        Parameters
+        ----------
+        x
+            `torch.Tensor` with leading batch dimension.
+
+        Returns
+        -------
+        `torch.Tensor` of ``bool`` values with leading batch dimension.
+
+        Raises
+        ------
+        ThresholdNotInferredError
+            If called before detector has had `infer_threshold` method called.
+        """
+        raw_scores = self.score(x)
+        scores = self._ensembler(raw_scores)
+        if not torch.jit.is_scripting():
+            self.check_threshold_inferred()
+        preds = scores > self.threshold
+        return preds
+
+    def _make_mask(self, reachabilities: torch.Tensor):
+        """Generate a mask for computing the average reachability.
+
+        If k is an array then we need to compute the average reachability for each k separately. To do
+        this we use a mask to weight the reachability of each k-close neighbor by 1/k and the rest to 0.
+        """
+        mask = torch.zeros_like(reachabilities[0], device=self.device)
+        for i, k in enumerate(self.ks):
+            mask[:k, i] = torch.ones(k, device=self.device)/k
+        return mask
+
+    def _compute_K(self, x, y):
+        """Compute the distance/similarity matrix matrix between `x` and `y`."""
```

Comment:
Could we remove "/similarity" here? A similarity matrix would have entried that icnrease with similarity, whereas this is the opposite